File: src/palettes/find-pane.ts
import { spawnSync } from "node:child_process"
import { definePalette } from "../palette"
import type { Item, RenderItemCtx } from "../types"

function tmux(args: string[]): string {
  const r = spawnSync("tmux", args, { stdio: ["ignore", "pipe", "pipe"] })
  return r.stdout?.toString().trimEnd() ?? ""
}

function detectAgent(command: string, title: string): string {
  const direct = new Set(["claude", "codex", "aider", "cursor-agent", "opencode", "gemini", "ollama"])
  if (direct.has(command)) return command
  if (title.startsWith("OC | ") || title.startsWith("OC|")) return "opencode"
  if (/^\s*[*✳⠂⠐⠁⠉⠙⠹⠸⠼⠴⠦⠧⠇⠏]\s/.test(title)) return "claude"
  return ""
}

type Pane = {
  session: string
  windowIndex: string
  paneIndex: string
  windowName: string
  paneTitle: string
  command: string
  path: string
  agent: string
  paneActive: boolean
  windowActive: boolean
  isCurrent: boolean
  target: string
}

type ItemData =
  | { kind: "session"; session: string; count: number; path: string; isCurrent: boolean }
  | { kind: "window"; session: string; windowIndex: string; windowName: string; treePrefix: string }
  | { kind: "pane"; pane: Pane; treePrefix: string }

function fetchPanes(): { panes: Pane[]; currentPane: string; currentSession: string } {
  const currentPane = tmux(["display-message", "-p", "#{session_name}:#{window_index}.#{pane_index}"])
  const currentSession = currentPane.split(":")[0] ?? ""

  const lines = tmux([
    "list-panes", "-a",
    "-F", [
      "#{session_name}",
      "#{window_index}",
      "#{pane_index}",
      "#{window_name}",
      "#{pane_title}",
      "#{pane_current_command}",
      "#{pane_current_path}",
      "#{pane_active}",
      "#{window_active}",
    ].join("\t"),
  ]).split("\n").filter(Boolean)

  const panes: Pane[] = []
  for (const line of lines) {
    const [session, windowIndex, paneIndex, windowName, paneTitle, command, path, paneActive, windowActive] = line.split("\t")
    if (!session || !windowIndex || !paneIndex) continue
    const target = `${session}:${windowIndex}.${paneIndex}`
    const title = paneTitle || `pane${paneIndex}`
    panes.push({
      session,
      windowIndex,
      paneIndex,
      windowName: windowName || `window${windowIndex}`,
      paneTitle: title,
      command: command || "",
      path: path || "",
      agent: detectAgent(command || "", title),
      paneActive: paneActive === "1",
      windowActive: windowActive === "1",
      isCurrent: target === currentPane,
      target,
    })
  }
  return { panes, currentPane, currentSession }
}

function buildItems(): Item[] {
  const { panes, currentSession } = fetchPanes()

  const sessionOrder: string[] = []
  const bySession = new Map<string, Map<string, { windowName: string; panes: Pane[] }>>()
  for (const p of panes) {
    if (!bySession.has(p.session)) {
      bySession.set(p.session, new Map())
      sessionOrder.push(p.session)
    }
    const ws = bySession.get(p.session)!
    if (!ws.has(p.windowIndex)) ws.set(p.windowIndex, { windowName: p.windowName, panes: [] })
    ws.get(p.windowIndex)!.panes.push(p)
  }

  const items: Item[] = []
  for (const session of sessionOrder) {
    const windows = [...bySession.get(session)!.entries()]
    const allInSession = windows.flatMap(([, w]) => w.panes)
    const focused = allInSession.find((p) => p.paneActive && p.windowActive) || allInSession[0]
    items.push({
      title: session,
      action: { tmux: `switch-client -t '${session}'` },
      selectable: false,
      data: {
        kind: "session",
        session,
        count: allInSession.length,
        path: focused?.path ?? "",
        isCurrent: session === currentSession,
      } satisfies ItemData,
    })

    for (let wi = 0; wi < windows.length; wi++) {
      const entry = windows[wi]!
      const [windowIndex, w] = entry
      const isLastWin = wi === windows.length - 1
      const winPrefix = `  ${isLastWin ? "└─" : "├─"} `

      if (w.panes.length === 1) {
        const p = w.panes[0]!
        items.push({
          title: p.paneTitle,
          action: {
            tmux: `select-pane -t '${p.target}' \\; select-window -t '${p.session}:${p.windowIndex}' \\; switch-client -t '${p.session}'`,
          },
          data: { kind: "pane", pane: p, treePrefix: winPrefix } satisfies ItemData,
        })
        continue
      }

      items.push({
        title: w.windowName,
        action: { tmux: `select-window -t '${session}:${windowIndex}' \\; switch-client -t '${session}'` },
        selectable: false,
        data: {
          kind: "window",
          session,
          windowIndex,
          windowName: w.windowName,
          treePrefix: winPrefix,
        } satisfies ItemData,
      })

      const panePrefixBase = isLastWin ? "      " : "  │   "
      for (let pi = 0; pi < w.panes.length; pi++) {
        const p = w.panes[pi]!
        const isLastPane = pi === w.panes.length - 1
        items.push({
          title: p.paneTitle,
          action: {
            tmux: `select-pane -t '${p.target}' \\; select-window -t '${p.session}:${p.windowIndex}' \\; switch-client -t '${p.session}'`,
          },
          data: {
            kind: "pane",
            pane: p,
            treePrefix: panePrefixBase + (isLastPane ? "└─ " : "├─ "),
          } satisfies ItemData,
        })
      }
    }
  }

  return items
}

function shortenPath(path: string): string {
  const home = process.env.HOME || ""
  return home && path.startsWith(home) ? `~${path.slice(home.length)}` : path
}

function renderItem(item: Item, ctx: RenderItemCtx): string {
  const { colors, active, width } = ctx
  const data = item.data as ItemData
  const rowBg = active ? colors.selected : colors.panel

  if (data.kind === "session") {
    const marker = data.isCurrent ? `${colors.accent}▶ ${colors.reset}${rowBg}` : "  "
    const name = `${colors.accent}${colors.bold}${data.session}${colors.reset}${rowBg}`
    const count = `${colors.muted} (${data.count})${colors.reset}${rowBg}`
    const path = data.path ? `  ${colors.muted}${shortenPath(data.path)}${colors.reset}${rowBg}` : ""
    return `${marker}${name}${count}${path}`
  }

  if (data.kind === "window") {
    const titleStyle = active ? colors.bold + colors.fg : colors.fg
    return `${colors.muted}${data.treePrefix}${colors.reset}${rowBg}${titleStyle}${data.windowName}${colors.reset}${rowBg}`
  }

  const p = data.pane
  let markerColor: string
  if (p.isCurrent) markerColor = colors.accent
  else if (p.paneActive) markerColor = "\x1b[38;2;166;227;161m"
  else markerColor = colors.muted
  const markerChar = p.isCurrent ? "▶" : p.paneActive ? "●" : "○"
  const titleStyle = p.isCurrent ? colors.muted : active ? colors.bold + colors.fg : colors.fg

  let left = `${colors.muted}${data.treePrefix}${colors.reset}${rowBg}${markerColor}${markerChar}${colors.reset}${rowBg} ${titleStyle}${p.paneTitle}${colors.reset}${rowBg}`
  let leftPlainW = data.treePrefix.length + 1 + 1 + p.paneTitle.length

  if (p.agent) {
    left += `  ${colors.muted}${p.agent}${colors.reset}${rowBg}`
    leftPlainW += 2 + p.agent.length
  }

  const rightText = `${p.windowIndex}.${p.paneIndex}`
  const right = `${colors.muted}${rightText}${colors.reset}${rowBg}`
  const gap = Math.max(1, width - leftPlainW - rightText.length)
  return `${left}${" ".repeat(gap)}${right}`
}

function filterTree(items: Item[], query: string): Item[] {
  const parts = query.toLowerCase().split(/\s+/).filter(Boolean)
  if (!parts.length) return items

  const okSessions = new Set<string>()
  const okWindows = new Set<string>()
  const okPanes = new Set<string>()

  for (const item of items) {
    const data = item.data as ItemData
    if (data.kind !== "pane") continue
    const p = data.pane
    const haystack = [
      p.session, p.windowName, p.paneTitle, p.command, p.path, p.target, p.agent,
    ].filter(Boolean).join(" ").toLowerCase()
    if (parts.every((part) => haystack.includes(part))) {
      okPanes.add(p.target)
      okSessions.add(p.session)
      okWindows.add(`${p.session}:${p.windowIndex}`)
    }
  }

  return items.filter((item) => {
    const data = item.data as ItemData
    if (data.kind === "session") return okSessions.has(data.session)
    if (data.kind === "window") return okWindows.has(`${data.session}:${data.windowIndex}`)
    return okPanes.has(data.pane.target)
  })
}

export const findPane = definePalette({
  title: "Find Pane",
  grouped: false,
  emptyText: "No panes",
  items: buildItems,
  renderItem,
  filter: filterTree,
})

File: crates/tmux-palette/src/palettes/find_pane.rs
use crate::fuzzy::multi_fuzzy_score;
use crate::model::{Action, Item, PaletteDef};
use crate::tmux::{tmux, tmux_quote};
use serde_json::json;

pub fn detect_agent(command: &str, title: &str) -> String {
    let direct = [
        "claude",
        "codex",
        "aider",
        "cursor-agent",
        "opencode",
        "gemini",
        "ollama",
    ];
    if direct.contains(&command) {
        return command.to_string();
    }
    if title.starts_with("OC | ") || title.starts_with("OC|") {
        return "opencode".into();
    }
    let trimmed_title = title.trim_start();
    if trimmed_title
        .chars()
        .next()
        .map(|c| "*✳⠂⠐⠁⠉⠙⠹⠸⠼⠴⠦⠧⠇⠏".contains(c))
        .unwrap_or(false)
        && trimmed_title.chars().nth(1) == Some(' ')
    {
        return "claude".into();
    }
    String::new()
}

#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Pane {
    pub session: String,
    pub window_index: String,
    pub pane_index: String,
    pub window_name: String,
    pub pane_title: String,
    pub command: String,
    pub path: String,
    pub agent: String,
    pub pane_active: bool,
    pub window_active: bool,
    pub is_current: bool,
    pub target: String,
}

pub fn parse_pane_line(line: &str, current_pane: &str) -> Option<Pane> {
    let parts: Vec<&str> = line.split('\t').collect();
    if parts.len() < 9 {
        return None;
    }
    if parts[0].is_empty() || parts[1].is_empty() || parts[2].is_empty() {
        return None;
    }
    let target = format!("{}:{}.{}", parts[0], parts[1], parts[2]);
    let pane_title = if parts[4].is_empty() {
        format!("pane{}", parts[2])
    } else {
        parts[4].to_string()
    };
    let command = parts[5].to_string();
    Some(Pane {
        session: parts[0].to_string(),
        window_index: parts[1].to_string(),
        pane_index: parts[2].to_string(),
        window_name: if parts[3].is_empty() {
            format!("window{}", parts[1])
        } else {
            parts[3].to_string()
        },
        pane_title: pane_title.clone(),
        command: command.clone(),
        path: parts[6].to_string(),
        agent: detect_agent(&command, &pane_title),
        pane_active: parts[7] == "1",
        window_active: parts[8] == "1",
        is_current: target == current_pane,
        target,
    })
}

pub fn pane_select_action(p: &Pane) -> Action {
    Action::Tmux {
        tmux: format!(
            "select-pane -t {} \\; select-window -t {} \\; switch-client -t {}",
            tmux_quote(&p.target),
            tmux_quote(&format!("{}:{}", p.session, p.window_index)),
            tmux_quote(&p.session)
        ),
    }
}

pub fn filter_tree(items: &[Item], query: &str) -> Vec<Item> {
    let parts: Vec<String> = query
        .to_lowercase()
        .split_whitespace()
        .map(|s| s.to_string())
        .collect();
    if parts.is_empty() {
        return items.to_vec();
    }
    let mut ok_sessions = std::collections::HashSet::new();
    let mut ok_windows = std::collections::HashSet::new();
    let mut ok_panes = std::collections::HashSet::new();
    for item in items {
        if let Some(data) = &item.data {
            if data.get("kind").and_then(|v| v.as_str()) == Some("pane") {
                let p = &data["pane"];
                let haystack = [
                    p.get("session"),
                    p.get("windowName"),
                    p.get("paneTitle"),
                    p.get("command"),
                    p.get("path"),
                    p.get("target"),
                    p.get("agent"),
                ]
                .into_iter()
                .filter_map(|v| v.and_then(|v| v.as_str()))
                .collect::<Vec<_>>()
                .join(" ");
                if multi_fuzzy_score(
                    &haystack,
                    &parts.iter().map(|s| s.as_str()).collect::<Vec<_>>(),
                ) > 0
                {
                    let session = p.get("session").and_then(|v| v.as_str()).unwrap_or("");
                    let window_index = p.get("windowIndex").and_then(|v| v.as_str()).unwrap_or("");
                    let target = p.get("target").and_then(|v| v.as_str()).unwrap_or("");
                    ok_panes.insert(target.to_string());
                    ok_sessions.insert(session.to_string());
                    ok_windows.insert(format!("{session}:{window_index}"));
                }
            }
        }
    }
    items
        .iter()
        .filter(|item| {
            if let Some(data) = &item.data {
                return match data.get("kind").and_then(|v| v.as_str()) {
                    Some("session") => ok_sessions
                        .contains(data.get("session").and_then(|v| v.as_str()).unwrap_or("")),
                    Some("window") => ok_windows.contains(&format!(
                        "{}:{}",
                        data.get("session").and_then(|v| v.as_str()).unwrap_or(""),
                        data.get("windowIndex")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                    )),
                    Some("pane") => ok_panes.contains(
                        data.get("pane")
                            .and_then(|v| v.get("target"))
                            .and_then(|v| v.as_str())
                            .unwrap_or(""),
                    ),
                    _ => false,
                };
            }
            false
        })
        .cloned()
        .collect()
}

fn build_items() -> Vec<Item> {
    let current_pane = tmux(&[
        "display-message",
        "-p",
        "#{session_name}:#{window_index}.#{pane_index}",
    ]);
    let current_session = current_pane.split(':').next().unwrap_or("").to_string();
    let lines: Vec<String> = tmux(&[
        "list-panes",
        "-a",
        "-F",
        "#{session_name}\t#{window_index}\t#{pane_index}\t#{window_name}\t#{pane_title}\t#{pane_current_command}\t#{pane_current_path}\t#{pane_active}\t#{window_active}",
    ])
    .lines()
    .filter(|s| !s.is_empty())
    .map(|s| s.to_string())
    .collect();
    let mut panes = Vec::new();
    for line in lines {
        if let Some(p) = parse_pane_line(&line, &current_pane) {
            panes.push(p);
        }
    }
    let mut items = Vec::new();
    let mut sessions = Vec::new();
    for p in &panes {
        if !sessions.contains(&p.session) {
            sessions.push(p.session.clone());
        }
    }
    for session in sessions {
        let in_session: Vec<&Pane> = panes.iter().filter(|p| p.session == session).collect();
        let focused = in_session
            .iter()
            .find(|p| p.pane_active && p.window_active)
            .copied()
            .or_else(|| in_session.first().copied());
        let mut session_item = Item::new(
            &session,
            Action::Tmux {
                tmux: format!("switch-client -t {}", tmux_quote(&session)),
            },
        );
        session_item.selectable = Some(false);
        session_item.data = Some(json!({
            "kind": "session",
            "session": session,
            "count": in_session.len(),
            "path": focused.map(|p| p.path.clone()).unwrap_or_default(),
            "isCurrent": session == current_session,
        }));
        items.push(session_item);
        for p in in_session {
            let mut pane_item = Item::new(&p.pane_title, pane_select_action(p));
            pane_item.data = Some(json!({"kind":"pane","pane": p}));
            items.push(pane_item);
        }
    }
    items
}

pub fn find_pane() -> PaletteDef {
    PaletteDef {
        title: Some("Find Pane".into()),
        items: build_items(),
        grouped: false,
        empty_text: Some("No panes".into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    #[test]
    fn detects_agents() {
        assert_eq!(detect_agent("claude", "x"), "claude");
        assert_eq!(detect_agent("sh", "OC | foo"), "opencode");
        assert_eq!(detect_agent("sh", "  ✳ working"), "claude");
    }
    #[test]
    fn parses_pane_line() {
        let p = parse_pane_line("s\t1\t2\tw\tt\tcmd\t/path\t1\t0", "s:1.2").unwrap();
        assert!(p.is_current);
        assert_eq!(p.agent, "");
    }

    #[test]
    fn rejects_incomplete_pane_lines() {
        assert!(parse_pane_line("\t1\t2\tw\tt\tcmd\t/path\t1\t0", "s:1.2").is_none());
    }
    #[test]
    fn selects_with_quoted_targets() {
        let p = Pane {
            session: "s".into(),
            window_index: "1".into(),
            pane_index: "2".into(),
            window_name: "w".into(),
            pane_title: "t".into(),
            command: "cmd".into(),
            path: "p".into(),
            agent: "".into(),
            pane_active: false,
            window_active: false,
            is_current: false,
            target: "s:1.2".into(),
        };
        if let Action::Tmux { tmux } = pane_select_action(&p) {
            assert!(tmux.contains("'s:1.2'"));
        } else {
            panic!();
        }
    }
    #[test]
    fn filters_tree_by_pane_matches() {
        let pane = json!({"kind":"pane","pane":{"session":"s","windowIndex":"1","paneIndex":"2","windowName":"w","paneTitle":"agent","command":"claude","path":"/x","agent":"claude","target":"s:1.2"}});
        let session = json!({"kind":"session","session":"s"});
        let item = Item {
            icon: None,
            icon_color: None,
            title: "x".into(),
            description: None,
            shortcut: None,
            category: None,
            aliases: vec![],
            action: Action::Palette {
                palette: "p".into(),
            },
            data: Some(pane),
            selectable: None,
        };
        let s = Item {
            data: Some(session),
            ..item.clone()
        };
        let filtered = filter_tree(&[s, item], "claude");
        assert_eq!(filtered.len(), 2);
    }
}
